pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -1178,12 +1190,46 @@ protected boolean verifyLocalBufferIsNotCorrupted(OpSendMsg op) {
         }
     }
 
+    static class ChunkedMessageCtx {
```

Comment:
It's not necessary to introduce this class as a field of `OpSendMsg` because `OpSendMsg` is already managed by Netty recycler. It's too complex to introduce a field that is also managed by Netty recycler.

I'd prefer adding two fields to `OpSendMsg`:

```java
        MessageIdImpl firstChunkMessageId;
        MessageIdImpl lastChunkMessageId;
```

Then in `ackReceived` you can just write

```java
        if (op.totalChunks > 1) {
            if (op.chunkId == 0) {
                op.firstChunkMessageId = new MessageIdImpl(ledgerId, entryId, partitionIndex);
            } else if (op.chunkId == op.totalChunks - 1) {
                op.lastChunkMessageId = new MessageIdImpl(ledgerId, entryId, partitionIndex);
                op.setMessageId(new ChunkMessageIdImpl(op.firstChunkMessageId, op.lastChunkMessageId));
            }
        }
```

After that, the code will be more clear.

Reply:
In addition, the `op.chunkedMessageCtx != null` check is redundant because `chunkedMessageCtx` is assigned when `totalChunks > 1`.

```java
                ChunkedMessageCtx chunkedMessageCtx = totalChunks > 1 ? ChunkedMessageCtx.get() : null;
```

Reply:
I think it's necessary. `OpSendMsg` is used to manage the publishing of a single chunk (or a single no-chunk message) but not the whole chunk message. When sending a chunk message, we need to create multiple OpSendMsg, but we only need one `ChunkedMessageCtx` to manage the context of all the chunks of that message.

Reply:
Sorry I missed it. You are right. Multiple `OpSendMsg`s share the same `ChunkedMessageCtx`.

However, I found a more serious problem and it should be a requested change. I found `ChunkedMessageCtx#recycle()` is never called. You should add some logic like reference count to call `recycle()` for this method.